File: src/main/java/vini2003/xyz/blade/Thing.java
package vini2003.xyz.blade;

import net.minecraft.network.PacketBuffer;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.network.NetworkEvent;

import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

public class Thing<T extends NetworkEvent> implements Consumer<T> {
    private final Class<T> clazz;
    private final Map<ResourceLocation, BiConsumer<NetworkEvent.Context, PacketBuffer>> map;
    
    public Thing(Class<T> clazz, Map<ResourceLocation, BiConsumer<NetworkEvent.Context, PacketBuffer>> map) {
        this.clazz = clazz;
        this.map = map;
    }
    
    @Override
    public void accept(T event) {
        if (event.getClass() == clazz) {
            NetworkEvent.Context context = event.getSource().get();
            
            if (!context.getPacketHandled()) {
                PacketBuffer buffer = new PacketBuffer(event.getPayload().copy());
                ResourceLocation type = buffer.readResourceLocation();
                BiConsumer<NetworkEvent.Context, PacketBuffer> consumer = map.get(type);
                
                if (consumer != null)
                    consumer.accept(context, buffer);
                
                context.setPacketHandled(true);
            }
        }
    }
}

File: src/main/kotlin/vini2003/xyz/blade/Thing.kt
package vini2003.xyz.blade

import net.minecraft.network.PacketBuffer
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.network.NetworkEvent
import java.util.function.BiConsumer
import java.util.function.Consumer
import kotlin.reflect.KClass

class Thing<T : NetworkEvent>(private val clazz: KClass<T>, private val map: Map<ResourceLocation, (NetworkEvent.Context, PacketBuffer) -> Unit>) : Consumer<T> {
    constructor(clazz: Class<T>, map: Map<ResourceLocation, BiConsumer<NetworkEvent.Context, PacketBuffer>>) : this(clazz.kotlin, map.convert())

    companion object {
        private fun Map<ResourceLocation, BiConsumer<NetworkEvent.Context, PacketBuffer>>.convert() =
            this.entries.map { it.key to { ctx: NetworkEvent.Context, buf: PacketBuffer -> it.value.accept(ctx, buf) } }.toMap()
    }

    override fun accept(event: T) {
        if (event::class != this.clazz) return

        val context = event.source.get()
        if (context.packetHandled) return

        val buffer = PacketBuffer(event.payload.copy())
        val type = buffer.readResourceLocation()

        this.map[type]?.let { it(context, buffer) }

        context.packetHandled = true
    }
}
